### PR TITLE
Introduce --config-file argument

### LINKED ISSUE #1369: Add `--config-file` CLI argument
### Current behavior:

I cannot specify a configuration file path

### Desired behavior:

I want to specify a configuration file path

#### Test code:

```
cypress run --config-file "/path/to/config"
```

### Additional Info (images, stack traces, etc)

I'm building a tool for PayPal called `paypal-scripts` (similar to `react-scripts`). One of the purposes is to cover 80% of the use cases by hiding all configuration from the user of the tool so they can focus on just writing the code. If they _do_ create a `cypress.json` file then we'll use that one, but for most people they should be able to just use the built-in configuration of `paypal-scripts` and they wont need to have and maintain a `cypress.json` file at all. There's honestly not a lot of configuration we'd need (mostly the `baseUrl` I think), but a lot of the value proposition of `paypal-scripts` is that people don't need to worry about configuring anything and it all just works out of the box.

I'm able to do this with `eslint`, `babel`, and many other tools, but I can't currently do it with `cypress` and I'd like to be able to do that if possible :) Thoughts?

### PR DESCRIPTION
**Pull request by flotwig**
Fixes #1369

### CODE REVIEW

File: cli/__snapshots__/cli_spec.js [hunk 1]
```diff
@@ -19,14 +19,15 @@ exports['shows help for open --foo 1'] = `
...
-    -c, --config <config>         sets configuration values. separate multiple values with a comma. overrides any value in cypress.json.
-    -d, --detached [bool]         runs Cypress application in detached mode
-    -P, --project <project path>  path to the project
-    --global                      force Cypress into global mode as if its globally installed
-    --dev                         runs cypress in development and bypasses binary check
-    -h, --help                    output usage information
+    -p, --port <port>                runs Cypress on a specific port. overrides any value in cypress.json.
+    -e, --env <env>                  sets environment variables. separate multiple values with a comma. overrides any value in cypress.json or cypress.env.json
+    -c, --config <config>            sets configuration values. separate multiple values with a comma. overrides any value in cypress.json.
+    -C, --config-file <config-file>  path to a config file to override the default cypress.json. set to "false" to skip specifying a config file.
```
Comment (jennifer-shehane) [inline cli/__snapshots__/cli_spec.js]: Is this required to be a JSON file? If so, this should be specified in description. Also would opt for use of word 'configuration' over 'config' in description as in previous description. Maybe:

"path to JSON file where configuration values are set. default path is './cypress.json'. pass 'false' to disable."

File: cli/lib/cli.js [hunk 2]
```diff
@@ -56,6 +56,7 @@ const descriptions = {
   detached: 'runs Cypress application in detached mode',
   project: 'path to the project',
   global: 'force Cypress into global mode as if its globally installed',
+  configFile: 'path to a config file to override the default cypress.json. set to "false" to skip specifying a config file.',
```
Comment (jennifer-shehane) [inline cli/lib/cli.js]: Is this required to be a JSON file? If so, this should be specified in description. Also would opt for use of word 'configuration' over 'config' in description as in previous description. Maybe:

"path to JSON file where configuration values are set. default path is './cypress.json'. pass 'false' to disable."

File: packages/desktop-gui/cypress/integration/settings_spec.coffee [hunk 3]
```diff
@@ -267,3 +269,21 @@ describe "Settings", ->
...
+      cy.contains("set from cypress.json (currently disabled by --config-file false)").should("exist")
+
+  context "when configFile is set", ->
+    beforeEach ->
+      @openProject.resolve(_.assign({ configFile: "special-cypress.json" }, @config))
+      @goToSettings()
+      cy.contains("Configuration").click()
+
+    it "notes that a custom config is in use", ->
+      cy.contains("set from custom config file special-cypress.json").should("exist")
```
Comment (jennifer-shehane) [inline packages/desktop-gui/cypress/integration/settings_spec.coffee]: Can we add a test where the user sets the config file to an invalid file type (if JSON is required) - say they set it to a `special-cypress.ts` file, for example - what happens?

File: packages/desktop-gui/src/runs/project-not-setup.jsx [hunk 4]
```diff
@@ -71,8 +72,8 @@ export default class ProjectNotSetup extends Component {
           <i className='fa fa-warning errored'></i>{' '}
           Runs cannot be displayed
         </h4>
-        <p>We were unable to find an existing project matching the <code>projectId</code> in your <code>cypress.json</code>.</p>
-        <p>To see runs for a current project, add the correct <code>projectId</code> to your <code>cypress.json</code></p>
+        <p>We were unable to find an existing project matching the <code>projectId</code> in your {configFileFormatted(this.props.project.configFile)}.</p>
+        <p>To see runs for a current project, add the correct <code>projectId</code> to your {configFileFormatted(this.props.project.configFile)}.</p>
```
Comment (jennifer-shehane) [inline packages/desktop-gui/src/runs/project-not-setup.jsx]: I would probably add the word 'file' to the end of these messages.

File: packages/server/lib/util/settings.js [hunk 5]
```diff
@@ -185,23 +202,27 @@ module.exports = {
...
 
-  remove (projectRoot) {
-    return fs.unlinkSync(this._pathToFile(projectRoot, 'cypress.json'))
+  remove (projectRoot, options = {}) {
+    return fs.unlinkSync(this.pathToCypressJson(projectRoot, options))
   },
 
-  pathToCypressJson (projectRoot) {
-    return this._pathToFile(projectRoot, 'cypress.json')
+  pathToCypressJson (projectRoot, options = {}) {
```
Comment (jennifer-shehane) [inline packages/server/lib/util/settings.js]: shouldn't this be named 'pathToConfigFile'?

File: packages/desktop-gui/cypress/integration/settings_spec.coffee (hunk 3 above)
Comment (flotwig) [reply packages/desktop-gui/cypress/integration/settings_spec.coffee]: If a user tries to use a file that doesn't contain JSON, the existing behavior still works (codeship.json has YAML in it):

image

The same error is thrown in run mode.

Comment (flotwig) [reply packages/desktop-gui/cypress/integration/settings_spec.coffee]: Should there be a filename restriction (only `.json` files) as well?

Comment (jennifer-shehane) [reply packages/desktop-gui/cypress/integration/settings_spec.coffee]: Yeah, what I was asking is, what happens if I pass `--configFile='foo.yml'`, not if I put invalid json data in a `.json` file?

Comment (flotwig) [reply packages/desktop-gui/cypress/integration/settings_spec.coffee]: It would try to read and write JSON from foo.yml, there's no filename validation there

Comment (flotwig) [reply packages/desktop-gui/cypress/integration/settings_spec.coffee]: Which is probably fine behavior, as it doesn't really matter what the extension is... they could be trying to use "--config-file cypress.json.test" to debug something, for example.

**Comment by Globegitter** [comment]
@jennifer-shehane what is the state of this? Are there still any blockers for getting this merged?

**Comment by Globegitter** [comment]
@jennifer-shehane another gentle ping.

**Comment by jennifer-shehane** [comment]
@Globegitter Thanks. Looking into this now.

**Comment by jennifer-shehane** [comment]
We need docs written for this: https://github.com/cypress-io/cypress-documentation/issues/1622

**Comment by Globegitter** [comment]
Does this need to be done before this is merged, or after this is merged and before a new version release?

**Comment by jennifer-shehane** [comment]
@Globegitter Before 3.3.0 release.

**Comment by jennifer-shehane** [comment]
### Passing a config-file without an extension. 

Error when no file matching, the "Could not find any tests to run" is a bit misleading ahead of the **actual** error.

Screen Shot 2019-04-26 at 1 45 42 PM

### Passing a config-file that is not JSON

We could probably have a better error display for this. I guess it does say it's trying to evaluate JSON....in a way, but maybe we could tell them it requires a `.json` file.

Screen Shot 2019-04-26 at 1 49 08 PM

Regardless, I think these are niceties - and this could likely go out without them, but it'd be best to handle them better now because we will probably forget to update it in the future.

**Comment by jennifer-shehane** [comment]
Oh, I didn't test passing this argument through the Module API and I also don't see any tests covering this use case. Is this supported? Cause I believe it should be.

**Comment by Toxicable** [comment]
Using this via the module api was my main use case for this
Ref: https://github.com/cypress-io/cypress/issues/3675

**Comment by Globegitter** [comment]
@flotwig gentle ping any chance to get the remaining issues cleared up?

**Comment by jennifer-shehane** [comment]
@Globegitter We discussed this with our team yesterday. This is @flotwig's next task after clearing up some 3.3.0 regressions.

**Comment by flotwig** [comment]
@Globegitter @jennifer-shehane I've added tests, documentation, and type defs for using this in the Module API. Works great from the Module API.

----

Re:

> Error when no file matching, the "Could not find any tests to run" is a bit misleading ahead of the **actual** error.
> Screen Shot 2019-04-26 at 1 45 42 PM

Yeah, this is weird, the error message is named `PROJECT_DOES_NOT_EXIST` but it's only called from `settings.js` when it can't find the configuration file:

https://github.com/cypress-io/cypress/blob/0f1f76286d5b2dd638c1ef5ebd03f5d5298d62e0/packages/server/lib/util/settings.js#L144-L147

So I've updated the error messaging and renamed it to `CONFIG_FILE_NOT_FOUND` internally to be a little clearer:

image

----

Re: only accepting .json files - I don't think it's necessary, the JSON parse error will be obvious enough, and we wouldn't want to prevent users from using a `cypress.json.prod` or `cypress.json.dev` naming scheme either.

----

Re: `--config-file` having different precedence than `cypress.json` - maybe I'm misunderstanding, but it's not possible to have the `cypress.env.json` override `baseUrl`, since it only applies to `env` variables, right?

I did some experiments and it seems to be working the same with cypress.json or with `--config-file`:

```js
// in cypress.json:
{ "env": { "test": "from cypress.json" } }
// in cypress.env.json
{ "env": { "test": "from cypress.env.json" } }
// in custom-config.json
{ "env": { "test": "from custom-config.json" } }
```

With `cypress open`, `cypress.env.json` overrides:

image

With `cypress open --config-file custom-config.json`, `cypress.env.json` still overrides:

image

Is there an issue here? If not, this is ready to be reviewed and merged.

File: cli/__snapshots__/cli_spec.js [hunk 6]
```diff
@@ -1,55 +1,33 @@
-exports['cli --version no binary version 1'] = `
```
Comment (Bkucera) [inline cli/__snapshots__/cli_spec.js]: any idea why these all got re-arranged? I think snap-shot-it should be sorting them in this version

I can't tell what changed as is

Comment (flotwig) [reply cli/__snapshots__/cli_spec.js]: any tips on getting them in the right order? I tried deleting `__snapshots__/cli_spec` and regenerating, but the diff is still meaningless compared to `develop`

**Comment by bahmutov** [comment]
Could you upgrade to latest snap-shot-it release please - it should keep the snapshots in the same order

**Comment by flotwig** [comment]
@bahmutov Doesn't seem to make a difference if I use 7.7.1 or 7.5.1

**Comment by bahmutov** [comment]
What I did to get snapshot `cli/__snapshots__/cli_spec.js` back to "stable" minimal diff

in folder `cli`:
- checked out the snapshot file from `develop` with `git checkout develop -- __snapshots__/cli_spec.js`
- ran just the `cli_spec.js` with 
```
BLUEBIRD_DEBUG=1 NODE_ENV=test ../node_modules/.bin/mocha --reporter mocha-multi-reporters --reporter-options configFile=../mocha-reporter-config.json ./test/lib/cli_spec.js
```
- there were just two changed snapshots due to the configuration file option. 
- **turned off whitespace trimming** in VSCode since our snapshots have whitespace in several lines :(
- **turned off autoformatting** in VSCode
- updated two snapshots with config file change
- reran test to confirm that new snapshots work and there is no pruning
- pushed the change - the GitHub diff shows only the 2 snapshots with changed text, all other snapshots stay in same place

File: packages/desktop-gui/src/project/project.jsx [hunk 7]
```diff
@@ -37,7 +37,7 @@ class Project extends Component {
   render () {
     if (this.props.project.isLoading) return <Loader color='#888' scale={0.5}/>
 
-    if (this.props.project.error) return <ErrorMessage error={this.props.project.error} onTryAgain={this._reopenProject}/>
+    if (this.props.project.error) return <ErrorMessage onTryAgain={this._reopenProject} project={this.props.project}/>
```
Comment (bahmutov) [inline packages/desktop-gui/src/project/project.jsx]: why not pass `error` any more?

File: packages/desktop-gui/src/runs/runs-list.jsx [hunk 8]
```diff
@@ -328,7 +329,7 @@ class RunsList extends Component {
           </h4>
           <h5>
             <span className='pull-left'>
-              1. Check <code>cypress.json</code> into source control.
+              1. Check {configFileFormatted(this.props.project.configFile)} into source control.
```
Comment (bahmutov) [inline packages/desktop-gui/src/runs/runs-list.jsx]: if `configFile=false` this message will be funky

File: packages/desktop-gui/src/settings/configuration.jsx [hunk 9]
```diff
@@ -130,11 +131,11 @@ const Configuration = observer(({ project }) => (
         </tr>
         <tr className='config-keys'>
           <td><span className='config'>config</span></td>
-          <td>set from cypress.json</td>
+          <td>set from {configFileFormatted(project.configFile)}</td>
```
Comment (bahmutov) [inline packages/desktop-gui/src/settings/configuration.jsx]: if `configFile=false` will be funky

Comment (bahmutov) [reply packages/desktop-gui/src/settings/configuration.jsx]: scratch that - you are returning a string that combines

File: packages/desktop-gui/src/runs/runs-list.jsx (hunk 8 above)
Comment (bahmutov) [reply packages/desktop-gui/src/runs/runs-list.jsx]: not most readable combination, but works

**Comment by bahmutov** [comment]
@flotwig could you please update the failing unit tests here

**Comment by flotwig** [comment]
@bahmutov 

> why do we have identical JSX file `const configFileFormatted = (configFile) => {` in runner and desktop-gui?

There's not a clean way to share code between the two packages. A cleaner workaround would be to create a new package to store shared resources but I figured it would be overkill for just this 1 function. I'm open to suggestions though.

File: packages/desktop-gui/src/project/project.jsx (hunk 7 above)
Comment (flotwig) [reply packages/desktop-gui/src/project/project.jsx]: `project` contains `error`, since we need to access more properties of `project` now I'm just passing `project` in directly

File: packages/desktop-gui/cypress/integration/settings_spec.coffee [hunk 10]
```diff
@@ -305,3 +307,21 @@ describe "Settings", ->
...
+      cy.contains("set from cypress.json file (currently disabled by --config-file false)").should("exist")
+
+  context "when configFile is set", ->
+    beforeEach ->
+      @openProject.resolve(_.assign({ configFile: "special-cypress.json" }, @config))
+      @goToSettings()
+      cy.contains("Configuration").click()
+
+    it "notes that a custom config is in use", ->
+      cy.contains("set from custom config file special-cypress.json").should("exist")
```
Comment (brian-mann) [inline packages/desktop-gui/cypress/integration/settings_spec.coffee]: `should('exist')` is the default assertion and is not necessary to write here

Comment (brian-mann) [inline packages/desktop-gui/cypress/integration/settings_spec.coffee]: should('exist') is the default assertion and is not necessary to write here

File: packages/server/__snapshots__/6_visit_spec.coffee.js [hunk 11]
```diff
@@ -507,7 +507,7 @@ exports['e2e visit normal response timeouts fails when visit times out 1'] = `
 
 Your page did not fire its 'load' event within '1000ms'.
 
-You can try increasing the 'pageLoadTimeout' value in 'cypress.json' to wait longer.
+You can try increasing the 'pageLoadTimeout' value in your configuration file to wait longer.
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_visit_spec.coffee.js]: I'm not sure how I feel about making these errors more generic. This leads to more "dead-end" behavior.

You did a lot of work in the **other** packages to provide the right value here, yet I feel like it's actually **more** important to do that when displaying errors... especially in `cypress run` mode.

File: packages/server/__snapshots__/6_visit_spec.coffee.js [hunk 12]
```diff
@@ -534,7 +534,7 @@ When this 'load' event occurs, Cypress will continue running commands.
 
 Your page did not fire its 'load' event within '500ms'.
 
-You can try increasing the 'pageLoadTimeout' value in 'cypress.json' to wait longer.
+You can try increasing the 'pageLoadTimeout' value in your configuration file to wait longer.
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_visit_spec.coffee.js]: see above comment ^^

File: packages/server/__snapshots__/6_web_security_spec.coffee.js [hunk 13]
```diff
@@ -46,7 +46,7 @@ Cypress does not allow you to change superdomains within a single test.
 
 You may need to restructure some of your test code to avoid this problem.
 
-Alternatively you can also disable Chrome Web Security which will turn off this restriction by setting { chromeWebSecurity: false } in your 'cypress.json' file.
+Alternatively you can also disable Chrome Web Security which will turn off this restriction by setting { chromeWebSecurity: false } in your configuration file.
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_web_security_spec.coffee.js]: see above comment ^^

File: packages/server/__snapshots__/6_web_security_spec.coffee.js [hunk 14]
```diff
@@ -78,7 +78,7 @@ Cypress does not allow you to change superdomains within a single test.
...
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_web_security_spec.coffee.js]: see above comment ^^

File: packages/server/__snapshots__/6_web_security_spec.coffee.js [hunk 15]
```diff
@@ -110,7 +110,7 @@ Cypress does not allow you to change superdomains within a single test.
...
```
Comment (brian-mann) [inline packages/server/__snapshots__/6_web_security_spec.coffee.js]: see above comment ^^

File: packages/server/lib/errors.coffee [hunk 16]
```diff
@@ -455,7 +455,7 @@ getMsgByType = (type, arg1 = {}, arg2) ->
       """
       We could not find a project with the ID: #{chalk.yellow(arg1)}
 
-      This projectId came from your cypress.json file or an environment variable.
+      This projectId came from your #{arg2} file or an environment variable.
```
Comment (brian-mann) [inline packages/server/lib/errors.coffee]: single quote `#{arg2}` here to be consistent

File: packages/server/lib/project.coffee [hunk 17]
```diff
@@ -488,7 +495,7 @@ class Project extends EE
     .map (projectRoot) ->
       Promise.props({
         path: projectRoot
-        id: settings.id(projectRoot)
+        id: settings.id(projectRoot, @options)
```
Comment (brian-mann) [inline packages/server/lib/project.coffee]: pretty sure `@options` is the wrong context here and will fail.

where's the associated test for this?

**Comment by bahmutov** [comment]
Probably should be “be.visible”?

File: packages/server/lib/util/settings.js [hunk 18]
```diff
@@ -117,11 +121,20 @@ module.exports = {
...
   },
 
-  exists (projectRoot) {
-    const file = this._pathToFile(projectRoot, 'cypress.json')
+  exists (projectRoot, options = {}) {
+    const file = this.pathToConfigFile(projectRoot, options)
 
     //# first check if cypress.json exists
-    return fs.statAsync(file)
+    return (
```
Comment (brian-mann) [inline packages/server/lib/util/settings.js]:
```js
// put this outside the `exists` function...
const maybeVerifyConfigFile = Promise.method((configFile) => {
  if (configFile === false) {
	return true
  }

  return fs.statAsync(configFile)
})

return maybeVerifyConfigFile(options.configFile)
.then(() => {
 ...
})
```

File: packages/server/lib/project.coffee (hunk 17 above)
Comment (flotwig) [reply packages/server/lib/project.coffee]: `@options` is set on `Project.open`, but there's no specific tests that it is factored in here besides these:

https://github.com/cypress-io/cypress/blob/2a5d88b641dd74079e2ffae1962994ac8f5a66d9/packages/server/test/unit/project_spec.coffee#L644-L663

I'll add some.

**Comment by TolikChirkov** [comment]
Gamble might McKenna is zhm grab

File: packages/server/lib/project.coffee (hunk 17 above)
Comment (TolikChirkov) [reply packages/server/lib/project.coffee]: Hadn't missed NZ vbsbkrm and it was

**Comment by TolikChirkov** [comment]
Jacket vbg is she doing #4615

File: packages/server/lib/project.coffee (hunk 17 above)
Comment (TolikChirkov) [reply packages/server/lib/project.coffee]: Ariana-grande-thank-u-next-lyrics
Ariana Grande – ​thank u, next Lyrics | Genius Lyrics

**Comment by brian-mann** [comment]
@flotwig 	Conflicts need to be resolved now

**Comment by Toxicable** [comment]
Is there anything else blocking this?

**Comment by pgroot91** [comment]
@Toxicable Would like to know that as well 👍

File: packages/server/lib/open_project.coffee [hunk 19]
```diff
@@ -201,6 +201,9 @@ create = ->
             relaunchBrowser()
       })
 
+      if !_.isUndefined(args.configFile)
+        options.configFile = args.configFile
```
Comment (brian-mann) [inline packages/server/lib/open_project.coffee]: this feels off to me... why is this necessary?

File: packages/server/lib/project.coffee (hunk 17 above)
Comment (brian-mann) [reply packages/server/lib/project.coffee]: i don't think this method is in use anyway - can you verify and remove the code + tests?

File: packages/server/lib/util/settings.js [hunk 20]
```diff
@@ -31,6 +31,14 @@ const flattenCypress = function (obj) {
   }
 }
 
+const maybeVerifyConfigFile = Promise.method((configFile) => {
+  if (configFile === false) {
+    return true
```
Comment (brian-mann) [inline packages/server/lib/util/settings.js]: think you can just `return` here, since the contract is not boolean based.

File: packages/server/lib/project.coffee (hunk 17 above)
Comment (flotwig) [reply packages/server/lib/project.coffee]: `desktop-gui` uses it: https://github.com/cypress-io/cypress/blob/774af917ce73d8129b8208a1d8c7b24f60e85557/packages/server/lib/gui/events.coffee#L166-L169

https://github.com/cypress-io/cypress/blob/774af917ce73d8129b8208a1d8c7b24f60e85557/packages/desktop-gui/src/projects/projects-api.js#L15-L46

Comment (flotwig) [reply packages/server/lib/project.coffee]: I think you're right @brian-mann, `@options` will be undefined since it's a static method. Now I'm wondering if more work actually needs to be done on this PR... this code is for the desktop-gui's project list. I guess up until now we could assume that the config file is always `cypress.json`, but it could be something else. The project cache stuff will need to be updated.

Comment (flotwig) [reply packages/server/lib/project.coffee]: There should probably also be a way for users in `open` mode to manually select and open a specific config file, to make this make sense.

Comment (flotwig) [reply packages/server/lib/project.coffee]: But then again, how would that work? If the configFile is in a subdirectory of the user's project, there's no way to reliably go from a `configFile` to a project path. 

So I'll do this:

* For now, there's no way to directly open a config file from the Cypress GUI
* Project openings with `--config-file` in GUI mode will be cached and will use the most recent `--config-file` upon being reopened from the projects list

Comment (flotwig) [reply packages/server/lib/project.coffee]: Okay, actually actually, projects are only cached in the projects list if they're opened directly from the file dialog in `cypress open`. So I think it's best to just assume that configFile is `cypress.json` here, and not cache projects if `configFile` isn't cypress.json for now.

If we want to add support for this later we can, but right now it seems like unnecessary functionality.

Comment (flotwig) [reply packages/server/lib/project.coffee]: 0441829927fde97bfa19b4bc338aebbe2cf7b1dd

File: packages/server/lib/open_project.coffee (hunk 19 above)
Comment (flotwig) [reply packages/server/lib/open_project.coffee]: so that the project gets initialized with the proper `configFile` option, if you remove it, the project won't use the correct `configFile`

**Comment by kentcdodds** [comment]
Yay!!! Toolkits rejoice!

**Comment by talentlessguy** [comment]
This is cool but it doesn't seem to be working :/

```
> yarn test:update
yarn run v1.17.3
$ cypress run --record --config-file update.cypress.json

  error: unknown option: --config-file
```

**Cypress version**: 3.4.1
